src/persist-client/src/write.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::marker::PhantomData;
+
+use differential_dataflow::difference::Semigroup;
+use mz_persist_types::{Codec, Codec64};
+use timely::progress::{Antichain, Timestamp};
+use tracing::warn;
+
+use crate::error::Error;
+
+pub struct WriteHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> WriteHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: Timestamp + Codec64,
+    D: Semigroup + Codec64,
+{
+    // This handle's upper frontier, not the global collection-level one.
+    pub fn upper(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
```

Comment:
nit: just to make sure I'm on the same page `new_upper` of the empty antichain finishes the collection for this particular write handle right? other write handles might still be writing stuff at various times?

Reply:
correct, though since they're required to be writing the same thing, presumably they'll follow suit shortly. because of that contract, it does mean we know everything that will ever be written to this collection tho (we don't really do anything with at atm, just pointing the implication out)